### PR TITLE
Support mysql ssl connection

### PR DESCRIPTION
**Pull request by bradymiller**
Am testing with the following mysql docker:
`docker run --detach --name mysql --env "MYSQL_ROOT_PASSWORD=root" -v ~/git/newcerts:/etc/newcertsfake:ro --net mynet mysql --character-set-server=utf8 --ssl-key=/etc/newcertsfake/server-key.pem --ssl-cert=/etc/newcertsfake/server-cert.pem --ssl-ca=/etc/newcertsfake/ca.pem`

Part of this issue:
https://github.com/openemr/openemr/issues/1113

demo is here:
http://www.open-emr.org/wiki/index.php/Development_Demo#Lambda_-_Up_For_Grabs_Demo

### CODE REVIEW

**Comment by bradymiller** [comment]
Now have support for the 3 adodb connections (main, gacl, calendar). Still need to do zend, new patient portal, and the new service thingy. Still not working very well (works at times and then doesn't work at times). Am hoping that when get all the connectors set up that it will work consistently.

File: gacl/gacl.class.php [hunk 1]
```diff
@@ -41,7 +41,11 @@
 
 //openemr configuration file - bm - 05-2009
 // to collect sql database login info and the utf8 flag
-include_once(dirname(__FILE__).'/../library/sqlconf.php');
+// also collect eh adodb libraries to support mysql_log that is needed for mysql ssl support
```
Comment (bradymiller) [inline gacl/gacl.class.php]: typos:
``````
// also collect the adodb libraries to support mysqli_mod that is needed for mysql ssl support

File: interface/main/calendar/config.php [hunk 2]
```diff
@@ -12,8 +12,15 @@
 //                  1 for encoded
 // ----------------------------------------------------------------------
 //
-include_once("../../../library/sqlconf.php");
-// Modified 5/2009 by BM for UTF-8 project 
+
+// to collect sql database login info and the utf8 flag
+// also collect eh adodb libraries to support mysql_log that is needed for mysql ssl support
```
Comment (bradymiller) [inline interface/main/calendar/config.php]: typo fix:
``````
// also collect the adodb libraries to support mysqli_mod that is needed for mysql ssl support

File: library/ADODB_mysqli_mod.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * ADODB custom wrapper class to support ssl option in gacl and calendar.
+ *
+ * @package   OpenEMR
+ * @link      http://www.open-emr.org
+ * @author    Sharon Cohen <sharonco@matrix.co.il>
+ * @author    Brady Miller <brady.g.miller@gmail.com>
+ * @copyright Copyright (c) 2017 Sharon Cohen <sharonco@matrix.co.il>
+ * @copyright Copyright (c) 2017 Brady Miller <brady.g.miller@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
```
Comment (bradymiller) [inline library/ADODB_mysqli_mod.php]: remove author/copyright Cohen

File: library/sql.inc [hunk 4]
```diff
@@ -123,18 +123,121 @@ class ADODB_mysqli_log extends ADODB_mysqli
 
         return $this->genID;
     }
+
```
Comment (bradymiller) [inline library/sql.inc]: Will migrate this custom ADODB_mysqli_log to a separate script called ADODB_mysqli_log.php

File: library/sql.inc [hunk 5]
```diff
@@ -123,18 +123,121 @@ class ADODB_mysqli_log extends ADODB_mysqli
...
+$database->optionFlags = array(array(MYSQLI_READ_DEFAULT_GROUP,0), array(MYSQLI_OPT_LOCAL_INFILE,1));
+// Set mysql to use ssl, if applicable.
+// Can support basic encryption by including just the mysql-ca pem (this is mandatory for ssl)
+// Can also support client based certificate if also include mysql-cert and mysql-key (this is optional for ssl)
+if (file_exists($GLOBALS['OE_SITE_DIR'] . "/documents/certificates/mysql-ca")) {
+    $database->clientFlags = MYSQLI_CLIENT_SSL;
+    error_log("DEBUG: Passing MYSQLI_CLIENT_SSL flag");
+}
 $database->port = $port;
 $database->PConnect($host, $login, $pass, $dbase);
```
Comment (bradymiller) [inline library/sql.inc]: Will likely need to change this to Connect (ie. not a persistent connection) to support ssl

File: gacl/gacl.class.php [hunk 6]
```diff
@@ -168,6 +172,14 @@ function __construct($options = NULL) {
...
+            // Set mysql to use ssl, if applicable.
+            // Can support basic encryption by including just the mysql-ca pem (this is mandatory for ssl)
+            // Can also support client based certificate if also include mysql-cert and mysql-key (this is optional for ssl)
+            if (file_exists($GLOBALS['OE_SITE_DIR'] . "/documents/certificates/mysql-ca")) {
+                $this->db->clientFlags = MYSQLI_CLIENT_SSL;
+                error_log("DEBUG: Passing MYSQLI_CLIENT_SSL flag from GACL");
+            }
+
 			// Port to be used in connection
 			$this->db->port = $sqlconf["port"];
```
Comment (bradymiller) [inline gacl/gacl.class.php]: At line 186, likely need to change to Connect (ie. remove persistent connection) to support ssl

**Comment by bradymiller** [comment]
@sjpadgett , just checking if the database connector here is being used in the patient portal:
https://github.com/openemr/openemr/blob/master/portal/patient/_machine_config.php
(I forgot whether you converted over to using the main openemr sql library or not)
thanks, -brady

**Comment by sjpadgett** [comment]
That's is my rest for patient documents and demographics(patient data). Every where else uses OpenEMR ado.

**Comment by bradymiller** [comment]
@sjpadgett , thanks for pointing me to the right places. I think I have a handle how I am going to squeeze out mysql ssl support out of the verysimple MySQLi class.

File: vendor/adodb/adodb-php/drivers/adodb-mysqli.inc.php [hunk 7]
```diff
@@ -107,6 +107,7 @@ function _connect($argHostname = NULL,
 		*/
 		foreach($this->optionFlags as $arr) {
 			mysqli_options($this->_connectionID,$arr[0],$arr[1]);
+			error_log("BAD WAY: ".$arr[0]." ".$arr[1]);
```
Comment (bradymiller) [inline vendor/adodb/adodb-php/drivers/adodb-mysqli.inc.php]: TODO:
I need to revert this script to original.

**Comment by bradymiller** [comment]
wooohoooo. Done and testing well!
Gonna bring this in very soon. Expediting this since we need to get support for this in 5.0.0 to ensure AWS packages that use separate Mysql servers are HIPAA eligible.

**Comment by bradymiller** [comment]
btw, see the README.md included for instructions on how to set up.

**Comment by bradymiller** [comment]
Doing well on extensive testing. Will bring into codebase.

**Pull request merged**